src/epoch_credits/mod.rs

```diff
@@ -1,35 +1,176 @@
-use solana_transaction_status::Rewards;
+use solana_runtime::bank::RewardType;
+use solana_sdk::{
+    clock::Epoch, epoch_info::EpochInfo, native_token::LAMPORTS_PER_SOL, pubkey::Pubkey,
+};
+use solana_stake_program::stake_state::StakeState;
+use solana_transaction_status::{Reward, Rewards};
+use std::{collections::BTreeSet, convert::TryFrom, u64};
 
 pub mod caching;
 pub mod caching_metadata;
 
 const SLOT_OFFSET: u64 = 20;
 
-pub fn get_rewards_for_epoch(epoch: Epoch, client: &RpcClient) -> anyhow::Result<Rewards> {
-    let info = client.get_epoch_info()?;
-
-    // Convert epoch number to slot
-    let start_slot = epoch * info.slots_in_epoch;
-
-    // We cannot use an excessively large range if the epoch just started. There is a chance that
-    // the end slot has not been reached and strange behaviour will occur.
-    // If this is the current epoch and less than `SLOT_OFFSET` slots have elapsed, then do not define an
-    // end_slot for use in the RPC call.
-    let end_slot = if info.epoch == epoch && info.slot_index < SLOT_OFFSET {
-        None
-    } else {
-        Some(start_slot + SLOT_OFFSET)
-    };
-
-    // First block only
-    let block = client
-        .get_blocks(start_slot, end_slot)?
-        .get(0)
-        .cloned()
-        .ok_or_else(|| anyhow!("no blocks found"))?;
-
-    Ok(client.get_block(block)?.rewards)
+struct StakingApy {
+    voter: String,
+    percent: f64,
+}
+
+struct ValidatorReward {
+    voter: String,
+    lamports: u64,
+}
+
+/// The monitor of rewards paid to validators and delegators.
+pub struct RewardsMonitor<'a> {
+    /// Shared Solana RPC client.
+    client: &'a RpcClient,
+    /// Prometheus staking APY gauge.
+    staking_apys: &'a GaugeVec,
+    /// Prometheus cumulative validator rewards gauge.
+    validator_rewards: &'a GaugeVec,
+    /// The last epoch in which rewards were recorded. 0 if uninitialised.
+    ///
+    /// FIXME: cached epoch credit history.
+    last_rewards_epoch: Epoch,
+}
+
+impl<'a> RewardsMonitor<'a> {
+    /// Initialises a new rewards monitor.
+    pub fn new(
+        client: &'a RpcClient,
+        staking_apys: &'a GaugeVec,
+        validator_rewards: &'a GaugeVec,
+    ) -> Self {
+        Self {
+            client,
+            staking_apys,
+            validator_rewards,
+            last_rewards_epoch: 0,
+        }
+    }
+
+    /// Exports reward metrics once an epoch.
+    pub fn export_rewards(&mut self, epoch_info: &EpochInfo) -> anyhow::Result<()> {
+        let epoch = epoch_info.epoch;
+        // FIXME: check seen epochs in the cache.
+        if epoch <= self.last_rewards_epoch {
+            return Ok(());
+        }
+        if let Some(rewards) = self.get_rewards_for_epoch(epoch, Some(epoch_info.clone()))? {
+            // FIXME: replace 3.0 with the previous epoch duration when using only one epoch, and
+            // with the average of all used epochs if using several.
+            let epochs_in_year = 365.0 / 3.0;
```

Comment:
Should this be dynamically calculated using data returned from the RPC (e.g. `slots in epoch * slot time`)? I am not sure if this is entirely possible since the RPC doesn't seem to have a way of exposing the time taken per slot.

Reply:
This is possible knowing the timestamps of the first blocks in the current and the previous epochs. Accuracy up to a few slots, i.e., within a few seconds. This should be easier to achieve with caching.

Reply:
Okay, I'll mark this as resolved and we'll come back to this after we get caching working!